### ISSUE TITLE
Optimize SAM

### ISSUE THREAD

**Issue by gabro** [scalacenter/scalafix#154]
Not sure this can be fully automated, but since Scala 2.12 has support for SAM some rewrite like this should be possible:

```scala
trait SomeSAM {
  def foo(x: Int)
}

val aSam = new SomeSAM {
  def foo(x: Int) = x
}
```

Can become (in Scala 2.12)

```scala
val aSam = (x: Int) => x
```

**Comment by olafurpg** [comment]
This is a cool idea! I think we will need `Type.members` to implement this rewrite because iiuc then the SAM type must have only one method?

> val aSam = (x: Int) => x

Should that be `val aSam: SomeSAM = (x: Int) => x`?

**Comment by gabro** [comment]
Correct, SAMs must have only one method (SAM stands for Single Abstract Method), and yes, I think the type annotation is required.

**Comment by gabro** [comment]
So a plan for this could be:

- look for instantiations of traits `new SomeTrait { ... }`
- check in the semanticdb that `SomeTrait` defines a single abstract method (not sure how doable this is in case of trait inheritance)
- replace the trait with a function (this is easy)

I'm not sure how many edge cases we're going to cover, but it should be ok, since we would probably not generating false positives. Simply, some patterns would go unrecognized and lead to empty patches. I think this is worth a shot.

**Comment by MasseGuillaume** [comment]
I'm starting to work on this.

The spec for the SAM conversion is defined here: http://www.scala-lang.org/files/archive/spec/2.12/06-expressions.html#sam-conversion

**Comment by joroKr21** [comment]
Does this rule exist somewhere now?

**Comment by bjaglin** [comment]
> Does this rule exist somewhere now?

Unfortunately, not to my knowledge.

**Comment by joroKr21** [comment]
> Unfortunately not, to my knowledge.

Thanks, I would like to have it so if I come up with something useful I will post back here 👍
